### PR TITLE
Add support for named parameters.

### PR DESCRIPTION
**Pull request by petermattis**
We don't currently have a way to send named parameters from the Go
driver and no way to specify them in the proto. The latter is easy to
fix, the former is tricky due to the interface mandated by
"database/sql".

### CODE REVIEW

File: sql/parser/expr.go [hunk 1]
```diff
@@ -232,10 +232,13 @@ func (node BoolVal) String() string {
 }
 
 // ValArg represents a named bind var argument.
-type ValArg int
+type ValArg string
 
 func (node ValArg) String() string {
-	return fmt.Sprintf("$%d", int(node))
+	if isDigit(int(node[0])) {
```
Comment (tamird) [inline sql/parser/expr.go]: why not use unicode.IsDigit here?

File: sql/parser/expr.go [hunk 2]
```diff
@@ -232,10 +232,13 @@ func (node BoolVal) String() string {
...
+		return fmt.Sprintf("$%s", string(node))
+	}
+	return fmt.Sprintf("@%s", string(node))
```
Comment (tamird) [inline sql/parser/expr.go]: `s/@/\$/`?

Comment (petermattis) [inline sql/parser/expr.go]: Oops, I was going to revert this to always use `$`. Forgot about that.

**Comment by tamird** [comment]
LGTM

**Comment by petermattis** [comment]
Uh, where did circleci go?

File: sql/parser/scan.go [hunk 3]
```diff
@@ -152,6 +152,17 @@ func (s *scanner) scan(lval *sqlSymType) {
 		if isDigit(s.peek()) {
 			s.scanParam(lval)
 			return
+		} else if s.syntax == Modern {
+			// TODO(pmattis): This should really be prefixed with '@', but that
```
Comment (spencerkimball) [inline sql/parser/scan.go]: So I assume that means we're going to change what we're doing for index indirection?

Comment (petermattis) [inline sql/parser/scan.go]: Yes. I'm not sure to what yet.

File: sql/server.go [hunk 4]
```diff
@@ -151,8 +153,17 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 type parameters []driver.Datum
 
 // Arg implements the Args interface
-func (p parameters) Arg(i int) (parser.Datum, bool) {
-	if i < 1 || i > len(p) {
+func (p parameters) Arg(name string) (parser.Datum, bool) {
+	if !unicode.IsDigit(rune(name[0])) {
```
Comment (bdarnell) [inline sql/server.go]: We can't use `unicode.IsDigit` here unless we also parse the number with a unicode-aware function. `strconv.ParseInt` only supports ASCII digits. `rune(name[0])` is also incorrect for non-ASCII characters; we must use functions in the `utf8` package instead.  I don't think we want to go down this path; even languages like Go that allow unicode identifiers still require that numbers use ASCII digits.

Comment (petermattis) [inline sql/server.go]: My thinking was that `rune(name[0])` would give the first byte as a `rune` which would only contain ASCII digits. I agree that is a bit subtle. I'll change this to something more explicit in a subsequent PR.

**Pull request merged**